# ---------------------------------------------------------------
# Author.... Merritt Burch
# Contact... mbb262@cornell.edu
# Date...... 2020-07-20 
#
# Description 
#   - Creating overlap file for askDB to create relationships
#   - between genes & intergenic regions to SNPs
# ---------------------------------------------------------------

# Load packages
library(data.table)
library(dplyr)
library(GenomicFeatures)
library(rtracklayer)

# List all files in directory
# scp -r mbb262@cbsublfs1.tc.cornell.edu:/data1/users/bm646/askdb_deployment_data /workdir/mbb262
setwd("/workdir/mbb262/askdb_deployment_data/")
my_files <- list.files(pattern = "vcf$")

# Get chromosome order
my_files_chrom <- gsub("sites_hmp321_agpv4_chr", "", my_files)
my_files_chrom <- as.numeric(as.character(gsub("\\.vcf", "", my_files_chrom)))


# ------------------
# Genic regions
# ------------------

# File directory
file_dir <- "/local/workdir/mbb262/askdb_deployment_data/"

# Lapply loop to find overlaps in genic regions, output them to new directory
lapply(seq_along(my_files),  function(i) {
  message("On file: ", my_files[i])
  
  # Load SNP sites taken from
  # scp -r mbb262@cbsublfs1.tc.cornell.edu:/data1/users/bm646/askdb_deployment_data/ /workdir/mbb262
  hap <- data.table::fread(paste0(file_dir, my_files[i]), select = 1:2)
  hap$V3 <- hap$V2
  names(hap) <- c("seqid", "start", "end")
  
  # Load in gene annotation file
  maize_v4 <- data.table::fread("maize_v4_genes.csv", select = c(2, 4, 5))
  maize_v4 <- maize_v4[seqid %in% my_files_chrom[i]] # Subset to a single chromosome

  # Set key and find overlaps between SNPs and genes
  setkey(hap, start, end)
  tmp <- foverlaps(maize_v4, hap, type="any")

  # Omit SNPs with no overlaps in genes regions
  tmp2 <- na.omit(tmp)

  # Rename columns and write to file
  names(tmp2) <- c("snp_chrom", "snp_start", "snp_end", "range_chrom", "range_start", "range_end")
  data.table::fwrite(tmp2, paste0("hapmap_", my_files_chrom[i], "_overlaps.csv"))

})


# ---------------------
# Intergenic regions
# ---------------------

# Get intergenic and genic regions from blfs1
# scp mbb262@cbsublfs1.tc.cornell.edu:/data1/users/mbb262/askdb_mgwas/genomic_range_data/Zea_mays.B73_RefGen_v4.47_gene_intergene_regions.csv /workdir/mbb262

# Change dir
setwd("/workdir/mbb262/overlap_files/")

# Load in file
intergene <- read.csv("/workdir/mbb262/Zea_mays.B73_RefGen_v4.47_gene_intergene_regions.csv", header = TRUE)
dim(intergene)

# Subset out only intergenic regions
intergene <- intergene %>% filter(type == "intergenic_region") %>% select("seqid", "start", "end", "intergenic_id")
intergene <- intergene %>% filter(!grepl('B73', seqid))
intergene <- intergene %>% filter(seqid != "Mt" & seqid != "Pt")
intergene$seqid <- as.numeric(as.character(intergene$seqid))
intergene <- intergene %>% data.table::as.data.table()
dim(intergene)
summary(intergene)

# Write intergenic file to disk
write.csv(intergene, "/workdir/mbb262/overlap_files/Zea_mays.B73_RefGen_v4.47_intergene_regions.csv", row.names = F, quote = F)

# Lapply loop to find overlaps in intergenic regions, output them to new directory
lapply(seq_along(my_files),  function(i) {
  message("On file: ", my_files[i])
  
  # Load SNP sites taken from
  # scp -r mbb262@cbsublfs1.tc.cornell.edu:/data1/users/bm646/askdb_deployment_data/ /workdir/mbb262
  hap <- data.table::fread(paste0(file_dir, my_files[i]), select = 1:2)
  hap$V3 <- hap$V2
  names(hap) <- c("seqid", "start", "end")
  
  # Load in intergenic annotation file
  maize_v4_intergene <- intergene[seqid %in% my_files_chrom[i]] # Subset to a single chromosome

  # Set key and find overlaps between SNPs and genes
  setkey(hap, start, end)
  tmp <- foverlaps(maize_v4_intergene, hap, type="any")

  # Omit SNPs with no overlaps in intergenic regions
  tmp2 <- na.omit(tmp)

  # Rename columns and write to file
  names(tmp2) <- c("snp_chrom", "snp_start", "snp_end", "range_chrom", "range_start", "range_end")
  data.table::fwrite(tmp2, paste0("hapmap_", my_files_chrom[i], "_intergenic_overlaps.csv"))

})


# ---------------------
#    Gene features
# ---------------------

# Load 'best' transcripts generated by Arcadio
# scp mbb262@cbsublfs1.tc.cornell.edu:/data1/users/mbb262/askdb_mgwas/genomic_range_data/NAM_alignments_allcds_byMaxScore4_5columns_bestB73transcripts_byQgene_withV4ID.txt /workdir/mbb262
v4_trans <- read.delim("/workdir/mbb262/NAM_alignments_allcds_byMaxScore4_5columns_bestB73transcripts_byQgene_withV4ID.txt")
v4_trans <- v4_trans %>% select("qnameV4") # subset to columns I need
v4_trans <-  v4_trans %>%  data.table::as.data.table()
v4_trans$qnameV4 <- as.character(v4_trans$qnameV4)

# Load complete gff file
# scp mbb262@cbsublfs1.tc.cornell.edu:/data1/users/mbb262/askdb_mgwas/genomic_range_data/Zea_mays.B73_RefGen_v4.47.gff3 /workdir/mbb262
whole_gff <- ape::read.gff("/workdir/mbb262/Zea_mays.B73_RefGen_v4.47.gff3", na.strings = c(".", "?"), GFF3 = TRUE)
whole_gff <- whole_gff %>% filter(seqid %in% c("1", "2", "3", "4", "5", "6", "7", "8", "9", "10")) # only take things from chroms 1:10

# Gather exons, UTRs
levels(whole_gff$type) # check levels
gff_trans <- whole_gff %>% filter(type %in% c("exon", "five_prime_UTR", "three_prime_UTR"))  # subset to CDS, exons, 

# Parse attribute names to transcript names and parent gene and exon/intron id column
gff_trans$transcript <- gsub(".*;Parent=transcript:", "", gff_trans$attributes)
gff_trans$transcript <- gsub("Parent=transcript:", "", gff_trans$transcript)
gff_trans$transcript <- gsub("ID=transcript:", "", gff_trans$transcript)
gff_trans$transcript <- gsub(";.*", "", gff_trans$transcript)
gff_trans <- gff_trans %>% data.table::as.data.table()

# Inner join with 'best' v4_transcripts
data.table::setkey(gff_trans, transcript)
data.table::setkey(v4_trans, qnameV4)
gff_trans <- gff_trans[v4_trans, nomatch = 0]
dim(gff_trans)

# Continue formatting
gff_trans$gene <- gsub("_T.*", "", gff_trans$transcript)
gff_trans$gene <- gsub("-T.*", "", gff_trans$gene)
gff_trans$exon_intron_id <- gsub(".*exon_id=", "", gff_trans$attributes)
gff_trans$exon_intron_id <- gsub(";.*", "", gff_trans$exon_intron_id)
gff_trans$exon_intron_id <- gsub("ID=transcript.*", "", gff_trans$exon_intron_id)
gff_trans$exon_intron_id <- gsub("Parent=transcript:.*", "", gff_trans$exon_intron_id)

# Add unique intron names
gff_trans$utr_id <- paste0(gff_trans$transcript, ".", gff_trans$type)
gff_trans$utr_id <- gsub(".*exon", "", gff_trans$utr_id) # remove rows where I pasted exon names

# Create introns
gtf <- GenomicFeatures::makeTxDbFromGFF(file = "/workdir/mbb262/Zea_mays.B73_RefGen_v4.47.gff3", format = "gff3")
introns <- GenomicFeatures::intronsByTranscript(gtf, use.names = TRUE) %>% 
  data.table::as.data.table() %>% 
  filter(seqnames %in% c("1", "2", "3", "4", "5", "6", "7", "8", "9", "10")) %>% 
  dplyr::select("seqnames", "start", "end", "group_name")

# Format transscript names, create gene names, do some formatting
introns$transcript <- gsub("transcript:", "", introns$group_name)
introns$gene <- gsub("_T", "", introns$transcript)
introns$gene <- gsub("-T", "", introns$gene)
colnames(introns)[1] <- "seqid" # change nanme of first coluymn to be consistent with gff file
introns <- introns %>% dplyr::select(-"group_name") # drop column

# only save transcripts matching 'good' transcripts
temp <- v4_trans$qnameV4
introns <- introns %>% filter(transcript %in% temp)
introns$type <- rep("intron", nrow(introns))

# create intron names
introns$exon_intron_id <- paste0(introns$transcript, ".intron", seq(1:nrow(introns)))

# Rbind fill rows together
together <- plyr::rbind.fill(gff_trans, introns) %>% arrange(seqid, start)

# Turn seqid numeric
together$seqid <- as.numeric(as.character(together$seqid))


# ---------------
# Export to file
# ---------------

# all
write.table(together, "/workdir/mbb262/intervals_all_mRNA_related_v4.csv", row.names = F, quote = F)

# exons & introns
temp <- together %>% 
  filter(type %in% c("exon", "intron")) %>% 
  dplyr::select("seqid", "start", "end", "exon_intron_id")
write.table(temp, "/workdir/mbb262/intervals_exons_introns_v4.csv", row.names = F, quote = F)

# UTRs
temp <- together %>% 
  filter(type %in% c("five_prime_UTR", "three_prime_UTR")) %>% 
  dplyr::select("seqid", "start", "end", "utr_id")
write.table(temp, "/workdir/mbb262/intervals_five_three_prime_utr_v4.csv", row.names = F, quote = F)


# -------------------------------------
# Find overlaps in exons, introns, UTRs
# -------------------------------------

# File directory
file_dir <- "/local/workdir/mbb262/askdb_deployment_data/"

# Exons & introns
lapply(seq_along(my_files),  function(i) {
  message("On file: ", my_files[i])
  
  # Load SNP sites taken from
  # scp -r mbb262@cbsublfs1.tc.cornell.edu:/data1/users/bm646/askdb_deployment_data/ /workdir/mbb262
  hap <- data.table::fread(paste0(file_dir, my_files[i]), select = 1:2)
  hap$V3 <- hap$V2
  names(hap) <- c("seqid", "start", "end")
  
  # Subset annotation file
  temp <- together %>% 
    filter(type %in% c("exon", "intron")) %>% 
    data.table::as.data.table() %>% 
    dplyr::select("seqid", "start", "end")
  maize_v4 <- temp[seqid %in% my_files_chrom[i]] # Subset to a single chromosome

  # Set key and find overlaps between SNPs and genes
  setkey(hap, start, end)
  tmp <- foverlaps(maize_v4, hap, type="any")
  
  # Omit SNPs with no overlaps in intergenic regions
  tmp2 <- na.omit(tmp)
  
  # Rename columns and write to file
  names(tmp2) <- c("snp_chrom", "snp_start", "snp_end", "range_chrom", "range_start", "range_end")
  data.table::fwrite(tmp2, paste0("hapmap_", my_files_chrom[i], "_exon_intron_overlaps.csv"))
  
})

# UTRs
lapply(seq_along(my_files),  function(i) {
  message("On file: ", my_files[i])
  
  # Load SNP sites taken from
  # scp -r mbb262@cbsublfs1.tc.cornell.edu:/data1/users/bm646/askdb_deployment_data/ /workdir/mbb262
  hap <- data.table::fread(paste0(file_dir, my_files[i]), select = 1:2)
  hap$V3 <- hap$V2
  names(hap) <- c("seqid", "start", "end")
  
  # Subset annotation file
  temp <- together %>% 
    filter(type %in% c("five_prime_UTR", "three_prime_UTR")) %>% 
    data.table::as.data.table() %>% 
    dplyr::select("seqid", "start", "end")
  maize_v4 <- temp[seqid %in% my_files_chrom[i]] # Subset to a single chromosome
  
  # Set key and find overlaps between SNPs and genes
  setkey(hap, start, end)
  tmp <- foverlaps(maize_v4, hap, type="any")
  
  # Omit SNPs with no overlaps in intergenic regions
  tmp2 <- na.omit(tmp)
  
  # Rename columns and write to file
  names(tmp2) <- c("snp_chrom", "snp_start", "snp_end", "range_chrom", "range_start", "range_end")
  data.table::fwrite(tmp2, paste0("hapmap_", my_files_chrom[i], "_five_three_prime_utr_overlaps.csv"))
  
})


# -----------------------
# Chromatin Status
# -----------------------

# Get chromatin status files from cbsu
# scp mbb262@cbsublfs1.tc.cornell.edu:/data1/users/mbb262/askdb_mgwas/genomic_range_data/AP.bfthresh1.1.MNaseHS.Ranges_v4_openclosedchromatin_intervals.csv /workdir/mbb262/askdb_deployment_data
# scp mbb262@cbsublfs1.tc.cornell.edu:/data1/users/mbb262/askdb_mgwas/genomic_range_data/RP.bfthresh1.1.MNaseHS.Ranges_v4_openclosedchromatin_intervals.csv /workdir/mbb262/askdb_deployment_data

# Load in files
ap <- read.csv("/workdir/mbb262/askdb_deployment_data/AP.bfthresh1.1.MNaseHS.Ranges_v4_openclosedchromatin_intervals.csv") %>% data.table::as.data.table()
rp <- read.csv("/workdir/mbb262/askdb_deployment_data/RP.bfthresh1.1.MNaseHS.Ranges_v4_openclosedchromatin_intervals.csv") %>% data.table::as.data.table()

# Lapply and find overlaps for ap
lapply(seq_along(my_files),  function(i) {
  message("On file: ", my_files[i])
  
  # Load SNP sites taken from
  # scp -r mbb262@cbsublfs1.tc.cornell.edu:/data1/users/bm646/askdb_deployment_data/ /workdir/mbb262
  hap <- data.table::fread(paste0(file_dir, my_files[i]), select = 1:2)
  hap$V3 <- hap$V2
  names(hap) <- c("seqid", "start", "end")
  
  # Subset annotation file
  temp <- ap %>% 
    dplyr::select("chrom", "start", "end")
  maize_v4 <- temp[chrom %in% my_files_chrom[i]] # Subset to a single chromosome
  
  # Set key and find overlaps between SNPs and genes
  setkey(hap, start, end)
  tmp <- foverlaps(maize_v4, hap, type="any")
  
  # Omit SNPs with no overlaps in intergenic regions
  tmp2 <- na.omit(tmp)
  
  # Rename columns and write to file
  names(tmp2) <- c("snp_chrom", "snp_start", "snp_end", "range_chrom", "range_start", "range_end")
  data.table::fwrite(tmp2, paste0("hapmap_", my_files_chrom[i], "_ap_openclosedchromatin_overlaps.csv"))
})

# Lapply and find overlaps for rp
lapply(seq_along(my_files),  function(i) {
  message("On file: ", my_files[i])
  
  # Load SNP sites taken from
  # scp -r mbb262@cbsublfs1.tc.cornell.edu:/data1/users/bm646/askdb_deployment_data/ /workdir/mbb262
  hap <- data.table::fread(paste0(file_dir, my_files[i]), select = 1:2)
  hap$V3 <- hap$V2
  names(hap) <- c("seqid", "start", "end")
  
  # Subset annotation file
  temp <- rp %>% 
    dplyr::select("chrom", "start", "end")
  maize_v4 <- temp[chrom %in% my_files_chrom[i]] # Subset to a single chromosome
  
  # Set key and find overlaps between SNPs and genes
  setkey(hap, start, end)
  tmp <- foverlaps(maize_v4, hap, type="any")
  
  # Omit SNPs with no overlaps in intergenic regions
  tmp2 <- na.omit(tmp)
  
  # Rename columns and write to file
  names(tmp2) <- c("snp_chrom", "snp_start", "snp_end", "range_chrom", "range_start", "range_end")
  data.table::fwrite(tmp2, paste0("hapmap_", my_files_chrom[i], "_rp_openclosedchromatin_overlaps.csv"))
})

